apps/rush-lib/src/logic/deploy/DeployManager.ts

```diff
@@ -0,0 +1,588 @@
+import ignore, { Ignore } from 'ignore';
+import {
+  Path,
+  FileSystem,
+  PackageJsonLookup,
+  FileSystemStats,
+  Sort,
+  JsonFile,
+  JsonSchema,
+  IPackageJson,
+  AlreadyExistsBehavior,
+  InternalError
+} from "@rushstack/node-core-library";
+import { RushConfiguration } from '../../api/RushConfiguration';
+import { SymlinkAnalyzer, ILinkInfo } from './SymlinkAnalyzer';
+import { RushConfigurationProject } from "../../api/RushConfigurationProject";
+
+// (@types/npm-packlist is missing this API)
+declare module "npm-packlist" {
+  export class WalkerSync {
+    public readonly result: string[];
+    public constructor(opts: { path: string });
+    public start(): void;
+  }
+}
+
+// Describes IDeployScenarioJson.projectSettings
+interface IDeployScenarioProjectJson {
+  projectName: string;
+  subdeploymentFolderName?: string;
+  additionalProjectsToInclude?: string[];
+}
+
+// The parsed JSON file structure, as defined by the "deploy-scenario.schema.json" JSON schema
+interface IDeployScenarioJson {
+  deploymentProjectNames: string[],
+  enableSubdeployments?: boolean;
+  includeDevDependencies?: boolean;
+  includeNpmIgnoreFiles?: boolean;
+  symlinkCreation?: "default" | "script" | "none";
+  projectSettings?: IDeployScenarioProjectJson[];
+}
+
+/**
+ * Stores additional information about folders being copied.
+ * Only some of the ISubdeploymentState.foldersToCopy items will an IFolderInfo object.
+ */
+interface IFolderInfo {
+  /**
+   * This is the lookup key for ISubdeploymentState.folderInfosByPath.
+   * It is an absolute real path.
+   */
+  folderPath: string;
+  /**
+   * True if this is the package folder for a local Rush project.
+   */
+  isRushProject: boolean;
+}
+
+/**
+ * This object tracks DeployManager state that is different for each subdeployment.
+ */
+interface ISubdeploymentState {
+  scenarioName: string;
+
+  mainProjectName: string;
+
+  /**
+   * The absolute path of the target folder for the subdeployment. If enableSubdeployments=false,
+   * then this points to the DeployManager._targetRootFolder.
+   */
+  targetSubdeploymentFolder: string;
+
+  /**
+   * During the analysis stage, _collectFoldersRecursive() uses this set to collect the absolute paths
+   * of the package folders to be copied.  The copying is performed later by _deployFolder().
+   */
+  foldersToCopy: Set<string>;
+
+  /**
+   * Additional information about some of the foldersToCopy paths.
+   * The key is the absolute real path from foldersToCopy.
+   */
+  folderInfosByPath: Map<string, IFolderInfo>;
+
+  symlinkAnalyzer: SymlinkAnalyzer;
+}
+
+/**
+ * Manages the business logic for the "rush deploy" command.
+ */
+export class DeployManager {
+  private static _jsonSchema: JsonSchema = JsonSchema.fromFile(
+    path.join(__dirname, '../../schemas/deploy-scenario.schema.json')
+  );
+
+  // Used by validateScenarioName()
+  private static _scenarioNameRegExp: RegExp = /^[a-z0-9]+(-[a-z0-9]+)*$/;
```

Comment:
Can you leave a comment summarizing what this matches?

Reply:
Fixed